--- chihuahua.c
+++ chihuahua.c
@@ -299,7 +299,7 @@
 }
 
 void collaps_sparsecnst(constraint *ocnst, statement *ost, const proof *pi, const sparsecnst *icnst, size_t k) {
-  size_t i,j,u,v;
+  size_t i,j,u,v,w;
   const size_t n = ost->n;
   int64_t s;
   polx (*phi)[n];
@@ -315,30 +315,32 @@
   shake128(hashbuf,sizeof(hashbuf),ost->h,16);
   memcpy(ost->h,hashbuf,16);
 
+  w = 0;
   for(i=0;i<k;i++) {
-    if(icnst->deg) continue;
+    const sparsecnst *cur = &icnst[i];
+    if(cur->deg) continue;
 
     s = 0;
     for(j=0;j<QBYTES;j++)
-      s |= (int64_t)hashbuf[16+i*QBYTES+j] << 8*j;
+      s |= (int64_t)hashbuf[16+w*QBYTES+j] << 8*j;
     s &= ((int64_t)1 << LOGQ) - 1;
+    w += 1;
 
-    if(icnst->b)
-      polx_scale_add(ocnst->b,icnst->b,s);
+    if(cur->b)
+      polx_scale_add(ocnst->b,cur->b,s);
 
-    for(j=0;j<icnst->nz;j++) {
+    for(j=0;j<cur->nz;j++) {
       phi = (polx(*)[n])ocnst->phi;
       u = v = 0;
-      while(u < icnst->idx[j]) {
+      while(u < cur->idx[j]) {
         phi += pi->nu[u];
         v = (pi->nu[u]) ? 0 : v + pi->n[u];
         u += 1;
       }
-      polxvec_scale_add(&(*phi)[v+icnst->off[j]],icnst->phi[j],icnst->len[j],s);
+      polxvec_scale_add(&(*phi)[v+cur->off[j]],cur->phi[j],cur->len[j],s);
     }
 
-    sparsemat_scale_add(ocnst->a,icnst->a,s);
-    icnst += 1;
+    sparsemat_scale_add(ocnst->a,cur->a,s);
   }
 }
 
@@ -352,27 +354,27 @@
   memcpy(ost->h,hashbuf,16);
 
   for(i=0;i<k;i++) {
-    if(cnst->deg == 0) continue;
-    polx alpha[cnst->deg];
-    polxvec_quarternary(alpha,cnst->deg,&hashbuf[16],i);
+    const sparsecnst *cur = &cnst[i];
+    if(cur->deg == 0) continue;
+    polx alpha[cur->deg];
+    polxvec_quarternary(alpha,cur->deg,&hashbuf[16],i);
 
-    if(cnst->b)
-      polxvec_sprod_add(ost->cnst->b,alpha,cnst->b,cnst->deg);
+    if(cur->b)
+      polxvec_sprod_add(ost->cnst->b,alpha,cur->b,cur->deg);
 
-    for(j=0;j<cnst->nz;j++) {
+    for(j=0;j<cur->nz;j++) {
       phi = (polx(*)[ost->n])ost->cnst->phi;
       u = v = 0;
-      while(u < cnst->idx[j]) {
+      while(u < cur->idx[j]) {
         phi += pi->nu[u];
         v = (pi->nu[u]) ? 0 : v + pi->n[u];
         u += 1;
       }
-      polxvec_collaps_add_extension(&(*phi)[v+cnst->off[j]],alpha,cnst->phi[j],cnst->len[j],
-                                    cnst->deg/cnst->mult[j],cnst->mult[j]);
+      polxvec_collaps_add_extension(&(*phi)[v+cur->off[j]],alpha,cur->phi[j],cur->len[j],
+                                    cur->deg/cur->mult[j],cur->mult[j]);
     }
 
-    sparsemat_polx_mul_add(ost->cnst->a,alpha,cnst->a);
-    cnst += 1;
+    sparsemat_polx_mul_add(ost->cnst->a,alpha,cur->a);
   }
 
   ost->cnst->a->coeffs = realloc(ost->cnst->a->coeffs,ost->cnst->a->len*sizeof(polx));
